feat(compliance): add FedRAMP 20x KSI compliance framework

ethanolivertroy opened the pull request:
### Context

This PR adds support for FedRAMP 20x Key Security Indicators (KSIs) compliance framework to Prowler. FedRAMP 20x is a modernization initiative aimed at automating the FedRAMP authorization process, focusing on continuous monitoring and cloud-native security principles. The 10 KSIs represent core security areas that cloud service providers must address as part of the FedRAMP 20x Phase One pilot program.

This framework enables organizations pursuing FedRAMP authorization to assess their cloud environments against the FedRAMP 20x requirements using Prowler's existing security checks.

### Description

This PR introduces FedRAMP 20x KSI compliance frameworks for AWS, Azure, and GCP providers. The implementation maps Prowler's existing security checks to the 10 Key Security Indicators defined by FedRAMP:

**Changes included:**
- Added 3 new compliance framework JSON files:
  - `prowler/compliance/aws/fedramp_20x_ksi_aws.json` - Maps 96 AWS checks to KSIs
  - `prowler/compliance/azure/fedramp_20x_ksi_azure.json` - Maps 73 Azure checks to KSIs  
  - `prowler/compliance/gcp/fedramp_20x_ksi_gcp.json` - Maps 94 GCP checks to KSIs

- Added dashboard visualization modules:
  - `dashboard/compliance/fedramp_20x_ksi_aws.py`
  - `dashboard/compliance/fedramp_20x_ksi_azure.py`
  - `dashboard/compliance/fedramp_20x_ksi_gcp.py`

- Updated documentation:
  - Updated framework counts in `docs/tutorials/compliance.md`

**The 10 KSIs covered:**
1. KSI-CED: Cybersecurity Education
2. KSI-CMT: Change Management
3. KSI-CNA: Cloud Native Architecture
4. KSI-IAM: Identity and Access Management
5. KSI-INR: Incident Reporting
6. KSI-MLA: Monitoring, Logging, and Auditing
7. KSI-PIY: Policy and Inventory
8. KSI-RPL: Recovery Planning
9. KSI-SVC: Service Configuration
10. KSI-TPR: Third-Party Information Resources

Each KSI is mapped to relevant NIST 800-53 controls and existing Prowler checks, following the official FedRAMP 20x documentation structure.

### Checklist

- Are there new checks included in this PR? **No** - This PR only adds compliance frameworks that map to existing checks
    - If so, do we need to update permissions for the provider? N/A
- [x] Review if the code is being covered by tests. *(Compliance frameworks follow existing patterns)*
- [x] Review if code is being documented following this specification https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
- [ ] Review if backport is needed.
- [ ] Review if is needed to change the [Readme.md](https://github.com/prowler-cloud/prowler/blob/master/README.md)
- [ ] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/prowler/CHANGELOG.md), if applicable.

#### API
- [ ] Verify if API specs need to be regenerated.
- [ ] Check if version updates are required (e.g., specs, Poetry, etc.).
- [ ] Ensure new entries are added to [CHANGELOG.md](https://github.com/prowler-cloud/prowler/blob/master/api/CHANGELOG.md), if applicable.

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

---

**Additional Notes for Reviewers:**
- Framework structure follows existing Prowler compliance framework patterns
- Check mappings are based on official FedRAMP 20x KSI documentation: https://github.com/FedRAMP/docs
- Tested locally with GCP provider, achieving expected compliance results
- Dashboard modules handle long KSI descriptions appropriately to prevent UI overlap issues

ethanolivertroy committed:
feat(aws): add Resource Control Policies (RCP) support and security checks

- Added RESOURCE_CONTROL_POLICY to the list of available organization policy types
- Added initial check to verify if RCPs are enabled and attached
- Added check for S3 security enforcement via RCPs
- Added NIST-aligned checks for encryption, IAM controls, logging/monitoring, and network security

ethanolivertroy committed:
fix: update ReadTheDocs configuration to use pip instead of Poetry

- Created docs/requirements.txt with required MkDocs dependencies
- Updated .readthedocs.yaml to use pip for dependency installation
- This should fix the ReadTheDocs build failure

ethanolivertroy committed:
Merge branch 'prowler-cloud:master' into master

ethanolivertroy committed:
Merge branch 'prowler-cloud:master' into master

ethanolivertroy committed:
Merge branch 'prowler-cloud:master' into master

ethanolivertroy committed:
feat: Add FedRAMP 20x Key Security Indicators (KSIs) compliance frameworks

- Add FedRAMP 20x KSI compliance framework for AWS
- Add FedRAMP 20x KSI compliance framework for Azure
- Add FedRAMP 20x KSI compliance framework for GCP
- Implement all 10 official FedRAMP KSIs based on FedRAMP documentation
- Update compliance documentation to include new frameworks
- Add comprehensive README for FedRAMP 20x KSI usage

The frameworks support FedRAMP 20x Phase One pilot requirements and focus on:
- Automation (80%+ automated validation target)
- Continuous monitoring
- Cloud-native security principles
- Zero trust architecture

Based on official FedRAMP KSI version 25.05C from:
https://github.com/FedRAMP/docs/blob/main/FRMR.KSI.key-security-indicators.json

ethanolivertroy committed:
feat: Add FedRAMP 20x Key Security Indicators (KSIs) compliance frameworks

- Add FedRAMP 20x KSI compliance framework for AWS
- Add FedRAMP 20x KSI compliance framework for Azure
- Add FedRAMP 20x KSI compliance framework for GCP
- Implement all 10 official FedRAMP KSIs based on FedRAMP documentation
- Update compliance documentation to include new frameworks
- Add comprehensive README for FedRAMP 20x KSI usage

The frameworks support FedRAMP 20x Phase One pilot requirements and focus on:
- Automation (80%+ automated validation target)
- Continuous monitoring
- Cloud-native security principles
- Zero trust architecture

Based on official FedRAMP KSI version 25.05C from:
https://github.com/FedRAMP/docs/blob/main/FRMR.KSI.key-security-indicators.json

ethanolivertroy committed:
feat: Add dashboard compliance modules for FedRAMP 20x KSI frameworks

- Add dashboard modules for AWS, Azure, and GCP FedRAMP 20x KSI display
- Fix text overlap issue by shortening long KSI descriptions in dashboard
- Enable proper visualization of FedRAMP 20x KSI compliance results

ethanolivertroy committed:
feat: Add dashboard compliance modules for FedRAMP 20x KSI frameworks

- Add dashboard modules for AWS, Azure, and GCP FedRAMP 20x KSI display
- Fix text overlap issue by shortening long KSI descriptions in dashboard
- Enable proper visualization of FedRAMP 20x KSI compliance results

ethanolivertroy commented:
<img width="2872" height="1994" alt="Screenshot 2025-08-12 at 09 19 41" src="https://github.com/user-attachments/assets/c368747c-2a3b-4dc6-8303-3ee70248c10a" />

<img width="2872" height="1994" alt="Screenshot 2025-08-12 at 09 22 22" src="https://github.com/user-attachments/assets/adeaf0d0-e336-4c20-83b9-15aaa6b8517a" />

ethanolivertroy committed:
Remove FEDRAMP_20X_KSI_README.md - not needed for main repo

ethanolivertroy committed:
Remove FEDRAMP_20X_KSI_README.md - not needed for main repo

ethanolivertroy committed:
test: Add comprehensive unit tests for FedRAMP 20x KSI compliance frameworks

- Add dashboard module tests for AWS, Azure, and GCP KSI visualizations
- Add compliance framework JSON validation tests
- Fix empty dataframe handling in dashboard modules
- Validate all 10 KSI requirements presence and structure
- Test NIST control mappings and version consistency
- Total: 29 unit tests covering all new functionality

Tests cover:
- Dashboard visualization functions (get_table)
- KSI description shortening logic
- Empty data edge cases
- Framework JSON structure validation
- Requirements completeness checks
- Cross-provider consistency

Addresses PR #8512 test requirements

ethanolivertroy committed:
test: Add comprehensive unit tests for FedRAMP 20x KSI compliance frameworks

- Add dashboard module tests for AWS, Azure, and GCP KSI visualizations
- Add compliance framework JSON validation tests
- Fix empty dataframe handling in dashboard modules
- Validate all 10 KSI requirements presence and structure
- Test NIST control mappings and version consistency
- Total: 29 unit tests covering all new functionality

Tests cover:
- Dashboard visualization functions (get_table)
- KSI description shortening logic
- Empty data edge cases
- Framework JSON structure validation
- Requirements completeness checks
- Cross-provider consistency

Addresses PR #8512 test requirements

ethanolivertroy committed:
Merge branch 'prowler-cloud:master' into feature/fedramp-20x-ksi-compliance

ethanolivertroy committed:
feat: Add FedRAMP 20x KSI compliance framework with tests

- Added FedRAMP 20x Key Security Indicators compliance frameworks for AWS, Azure, and GCP
- Included comprehensive unit tests for all frameworks
- Updated CHANGELOG.md with new feature entry
- Fixed missing newlines at end of files per pre-commit hooks
- Removed RCP (Resource Control Policy) features that belong in separate branch

ethanolivertroy committed:
feat: Add FedRAMP 20x KSI compliance framework with tests

- Added FedRAMP 20x Key Security Indicators compliance frameworks for AWS, Azure, and GCP
- Included comprehensive unit tests for all frameworks
- Updated CHANGELOG.md with new feature entry
- Fixed missing newlines at end of files per pre-commit hooks
- Removed RCP (Resource Control Policy) features that belong in separate branch

MrCloudSec committed:
chore: revision

ethanolivertroy committed:
fix(tests): correct FedRAMP KSI identifiers to match official schema

Fixed incorrect KSI identifier codes in dashboard test files to align with
official FedRAMP 20x KSI schema and our compliance framework files:
- Changed ksi-mon to ksi-mla (Monitoring, Logging, and Auditing)
- Changed ksi-pol to ksi-piy (Policy and Inventory)
- Changed ksi-rec to ksi-rpl (Recovery Planning)

This resolves the issues identified by GitHub Copilot in PR #8512.

ethanolivertroy committed:
fix(tests): correct FedRAMP KSI identifiers to match official schema

Fixed incorrect KSI identifier codes in dashboard test files to align with
official FedRAMP 20x KSI schema and our compliance framework files:
- Changed ksi-mon to ksi-mla (Monitoring, Logging, and Auditing)
- Changed ksi-pol to ksi-piy (Policy and Inventory)
- Changed ksi-rec to ksi-rpl (Recovery Planning)

This resolves the issues identified by GitHub Copilot in PR #8512.

ethanolivertroy committed:
fix(tests): complete FedRAMP KSI identifier fixes from code review

- Fixed ksi-inc to ksi-inr in Azure test file
- Removed hardcoded string slice in dashboard modules for accurate matching
- Aligns with official FedRAMP 20x KSI schema

Addresses all issues from GitHub Copilot review in PR #8512

ethanolivertroy committed:
fix(tests): complete FedRAMP KSI identifier fixes from code review

- Fixed ksi-inc to ksi-inr in Azure test file
- Removed hardcoded string slice in dashboard modules for accurate matching
- Aligns with official FedRAMP 20x KSI schema

Addresses all issues from GitHub Copilot review in PR #8512

pedrooot reviewed (changes requested):
Thanks for this, please review my comments and remember to add the needed changes in `__main__.py` in order to generate the output for the new compliance frameworks that you're adding. See how it's done here: https://github.com/prowler-cloud/prowler/blob/bb07cf9147cde9b911273dc1b5518d28f677a992/prowler/__main__.py#L436C13-L448C47

ethanolivertroy committed:
chore: merge latest master into feature branch

ethanolivertroy committed:
chore: resolve merge conflicts from stashed changes

ethanolivertroy committed:
fix(compliance): rename FedRAMP 20x KSI to Low and remove NistControls

BREAKING CHANGE: FedRAMP 20x KSI framework files renamed to include "_low" suffix

Changes made to address PR #8512 feedback:
- Removed custom NistControls attribute from all FedRAMP 20x KSI JSON files
- Renamed all fedramp_20x_ksi files to fedramp_20x_ksi_low to distinguish from upcoming moderate level
- Updated framework names to "FedRAMP-20x-KSI-Low"
- Updated all test imports and file references
- Removed test for NistControls attribute since it's no longer needed
- Frameworks now use generic compliance handler (no __main__.py changes needed)

Files renamed:
- prowler/compliance/{aws,azure,gcp}/fedramp_20x_ksi_*.json → fedramp_20x_ksi_low_*.json
- dashboard/compliance/fedramp_20x_ksi_*.py → fedramp_20x_ksi_low_*.py
- tests files similarly updated

This prepares the codebase for adding FedRAMP 20x KSI Moderate level frameworks
in accordance with RFC-0014 Phase Two requirements.

Resolves feedback from @pedrooot on PR #8512

ethanolivertroy committed:
fix(compliance): rename FedRAMP 20x KSI to Low and remove NistControls

BREAKING CHANGE: FedRAMP 20x KSI framework files renamed to include "_low" suffix

Changes made to address PR #8512 feedback:
- Removed custom NistControls attribute from all FedRAMP 20x KSI JSON files
- Renamed all fedramp_20x_ksi files to fedramp_20x_ksi_low to distinguish from upcoming moderate level
- Updated framework names to "FedRAMP-20x-KSI-Low"
- Updated all test imports and file references
- Removed test for NistControls attribute since it's no longer needed
- Frameworks now use generic compliance handler (no __main__.py changes needed)

Files renamed:
- prowler/compliance/{aws,azure,gcp}/fedramp_20x_ksi_*.json → fedramp_20x_ksi_low_*.json
- dashboard/compliance/fedramp_20x_ksi_*.py → fedramp_20x_ksi_low_*.py
- tests files similarly updated

This prepares the codebase for adding FedRAMP 20x KSI Moderate level frameworks
in accordance with RFC-0014 Phase Two requirements.

Resolves feedback from @pedrooot on PR #8512

ethanolivertroy committed:
feat: Add RFC-0014 Phase Two KSIs to FedRAMP 20x Low frameworks

Implements FedRAMP RFC-0014 Phase Two requirements for Low impact level:

Changes:
- Renamed KSI-INR from "Incident Reporting" to "Incident Response" per RFC-0014
- Added 3 new Low KSIs (13 total, up from 10):
  * KSI-CED-03: Role-specific training for dev/engineering staff
  * KSI-IAM-07: Account lifecycle and privilege management
  * KSI-MLA-07: Monitoring and logging inventory
- Updated tests to validate 13 KSIs instead of 10

Coverage updates:
- AWS: 175 checks (13 KSIs)
- Azure: 186 checks (13 KSIs)
- GCP: 122 checks (13 KSIs)

All framework tests passing.

ethanolivertroy committed:
feat: Add RFC-0014 Phase Two KSIs to FedRAMP 20x Low frameworks

Implements FedRAMP RFC-0014 Phase Two requirements for Low impact level:

Changes:
- Renamed KSI-INR from "Incident Reporting" to "Incident Response" per RFC-0014
- Added 3 new Low KSIs (13 total, up from 10):
  * KSI-CED-03: Role-specific training for dev/engineering staff
  * KSI-IAM-07: Account lifecycle and privilege management
  * KSI-MLA-07: Monitoring and logging inventory
- Updated tests to validate 13 KSIs instead of 10

Coverage updates:
- AWS: 175 checks (13 KSIs)
- Azure: 186 checks (13 KSIs)
- GCP: 122 checks (13 KSIs)

All framework tests passing.

ethanolivertroy committed:
fix: add required Name field and RFC-0014 Phase Two KSIs to FedRAMP 20x Low frameworks

This commit addresses framework validation issues and implements FedRAMP RFC-0014
Phase Two requirements for Low impact level compliance frameworks.

Changes:
- Added required 'Name' field to framework metadata for all three providers
  * Enables frameworks to appear in --list-compliance output
  * Required for Prowler compliance framework validation
  * Fixes: 'field required (type=value_error.missing)' validation error

- Renamed KSI-INR from 'Incident Reporting' to 'Incident Response' per RFC-0014

- Added 3 new Phase Two Low KSIs (total now 13 instead of 10):
  * KSI-CED-03: Role-specific training for dev/engineering staff
  * KSI-IAM-07: Account lifecycle and privilege management
  * KSI-MLA-07: Monitoring and logging inventory

Framework Coverage:
- AWS: 175 checks across 13 KSIs
- Azure: 186 checks across 13 KSIs
- GCP: 122 checks across 13 KSIs

Testing:
- All framework unit tests passing (13/13)
- Successfully tested with live AWS account scan (614 findings)
- Dashboard visualization confirmed working
- Frameworks load correctly in Prowler CLI and web dashboard

Compliance: Fully aligned with FedRAMP RFC-0014 Phase Two specification

ethanolivertroy committed:
fix: add required Name field and RFC-0014 Phase Two KSIs to FedRAMP 20x Low frameworks

This commit addresses framework validation issues and implements FedRAMP RFC-0014
Phase Two requirements for Low impact level compliance frameworks.

Changes:
- Added required 'Name' field to framework metadata for all three providers
  * Enables frameworks to appear in --list-compliance output
  * Required for Prowler compliance framework validation
  * Fixes: 'field required (type=value_error.missing)' validation error

- Renamed KSI-INR from 'Incident Reporting' to 'Incident Response' per RFC-0014

- Added 3 new Phase Two Low KSIs (total now 13 instead of 10):
  * KSI-CED-03: Role-specific training for dev/engineering staff
  * KSI-IAM-07: Account lifecycle and privilege management
  * KSI-MLA-07: Monitoring and logging inventory

Framework Coverage:
- AWS: 175 checks across 13 KSIs
- Azure: 186 checks across 13 KSIs
- GCP: 122 checks across 13 KSIs

Testing:
- All framework unit tests passing (13/13)
- Successfully tested with live AWS account scan (614 findings)
- Dashboard visualization confirmed working
- Frameworks load correctly in Prowler CLI and web dashboard

Compliance: Fully aligned with FedRAMP RFC-0014 Phase Two specification

ethanolivertroy commented on prowler/compliance/aws/fedramp_20x_ksi_aws.json:
```diff
@@ -0,0 +1,310 @@
+        {
+          "ItemId": "ksi-ced",
+          "Section": "Cybersecurity Education",
+          "NistControls": "at-2, at-2.2, at-2.3, at-4",
```
resolved in latest commit

ethanolivertroy commented:
<img width="1798" height="1303" alt="CleanShot 2025-10-29 at 11 52 15" src="https://github.com/user-attachments/assets/0e7016bf-8497-415c-a39d-5da93a985d0f" />

Updated for recent KSI changes in RFC - https://www.fedramp.gov/rfcs/0014/

ethanolivertroy commented:
https://github.com/user-attachments/assets/20d8d84c-8da6-4831-a327-f438d17f3a8b

Tested this with am AWS test environment.

ethanolivertroy commented:
@pedrooot let me know what you think when you get a chance, I'll submit a separate PR for the moderate KSIs, these are just the low impact ones in this PR

pedrooot committed:
chore(revision): add some changes

pedrooot committed:
chore(revision): add some changes

ethanolivertroy commented:
@pedrooot thanks for removing that CED Education check, I was trying to make it work but makes sense to just leave it out for now

I removed the unneeded changes inside of API, sorry think I got some branch work mixed up 😅 

Dropping this here as a little disclaimer. There is a current government shutdown so these are still in draft.
I will update as they finalize though as this is still useful for orgs trying to self gap assess themselves now.

# FedRAMP 20x KSI Low Compliance Framework - Limitations & Usage Guide

## Overview

The FedRAMP 20x Key Security Indicators (KSI) Low Impact Level compliance frameworks provide automated security assessment for cloud service providers pursuing FedRAMP Low authorization. These frameworks implement [RFC-0014 Phase Two](https://www.fedramp.gov/rfcs/0014/) requirements across AWS, Azure, and GCP.

## Scope & Purpose

**What This Framework Checks:**
- Technical cloud infrastructure security controls
- Identity and access management configurations
- Encryption implementation (at rest and in transit)
- Network segmentation and traffic controls
- Logging, monitoring, and audit configurations
- Service configuration security settings
- Change management through automated tooling
- High availability and resilience architecture

**What This Framework Does NOT Check:**
- Organizational policies and procedures
- Employee training and awareness programs
- Physical security controls
- Manual documentation reviews
- Contractual agreements with third parties
- Business continuity plan documents
- Incident response runbooks (documents vs. technical controls)

## Known Limitations

### 1. Organizational & Training KSIs Not Covered

The following RFC-0014 Phase Two KSIs are **intentionally excluded** because they require organizational policy validation rather than technical infrastructure checks:

- **KSI-CED-01** (Security & Privacy Awareness Training): Requires evidence of employee training programs, completion rates, and testing results
- **KSI-CED-03** (Role-Specific Security Training): Requires documentation of role-based training curricula and completion tracking

**Rationale:** Prowler focuses on automated technical security assessments of cloud infrastructure. These education requirements must be validated through:
- Policy document review
- Training management system reports
- HR records and completion certificates
- Manual interviews and audits

### 2. Technical Implementation Approach

This framework uses an **umbrella category approach** to organize checks:

| Umbrella Category | RFC-0014 Phase Two Requirements Covered |
|-------------------|----------------------------------------|
| **KSI-CMT** (Change Management) | CMT-01 (Service Modification Logging), CMT-04 (Change Management Compliance) |
| **KSI-CNA** (Cloud Native Architecture) | CNA-01 (Network Traffic Limitation) |
| **KSI-IAM** (Identity & Access Management) | IAM-05 (Zero Trust Design) |
| **KSI-SVC** (Service Configuration) | SVC-03 (Encryption at Rest) |
| **KSI-IAM-07** (Account Lifecycle) | Explicitly implemented as dedicated requirement |
| **KSI-MLA-07** (Monitoring Inventory) | Explicitly implemented as dedicated requirement |

**Benefits of Umbrella Approach:**
- Comprehensive multi-layered coverage (defense in depth)
- Easier maintenance across 483+ checks
- Cloud-native service focus
- Practical implementation for CSP validation

### 3. CI/CD Pipeline Security Gap

**RFC-0014 Requirement:** KSI-CMT-03 (Persistent Automated Testing)

**Current Coverage:** Partial
- Automated compliance monitoring (AWS Config, SSM)
- Automated patch management
- **Gap:** CI/CD pipeline security checks

**Missing Technical Checks:**
- CodeBuild project security configurations
- CodePipeline artifact encryption
- Build-time secret management
- Container image scanning in pipelines

**Why This Gap Exists:**
- Many CSPs use external CI/CD tools (Jenkins, GitLab CI, GitHub Actions) rather than AWS-native services
- CI/CD security implementation varies significantly by organization
- The framework focuses on deployed infrastructure rather than build pipelines

**Mitigation:** Organizations using AWS CodeBuild/CodePipeline should manually validate:
- Artifact encryption is enabled
- Build logs are configured and retained
- Secrets are managed through AWS Secrets Manager or Parameter Store
- Pipeline webhooks use authentication
- Build environments use least privilege IAM roles

### 4. Subject to RFC-0014 Finalization

**Current Implementation:** Based on RFC-0014 Phase Two as published at https://www.fedramp.gov/rfcs/0014/

**Important Notes:**
- RFC-0014 Phase Two is an **official FedRAMP publication** but may receive updates
- This framework implements the specification as of **November 2025**
- Framework will be updated if FedRAMP publishes material changes to KSI requirements
- Users should monitor https://www.fedramp.gov/rfcs/ for official updates

**Moderate Impact Level:** RFC-0014 Phase Two also includes KSIs for Moderate impact level (KSI-CNA-08, KSI-MLA-08, KSI-SVC-08, KSI-SVC-09, KSI-SVC-10). These are **not included** in this Low framework and will be addressed in a future implementation.

## Coverage Analysis

### RFC-0014 Phase Two Low Impact Level Requirements

| Requirement | Coverage | Check Count | Status |
|-------------|----------|-------------|--------|
| KSI-CMT-01 (Service Modification Logging) | Excellent | 15 | PASS |
| KSI-CMT-03 (Persistent Automated Testing) | Partial | 2-3 | PARTIAL |
| KSI-CMT-04 (Change Management Compliance) | Good | 6 | PASS |
| KSI-CNA-01 (Network Traffic Limitation) | Excellent | 9 | PASS |
| KSI-IAM-05 (Zero Trust Design) | Excellent | 12 | PASS |
| KSI-IAM-07 (Account Lifecycle Management) | Explicit | 6 | PASS |
| KSI-MLA-07 (Monitoring & Logging Inventory) | Explicit | 5 | PASS |
| KSI-SVC-03 (Encryption at Rest) | Excellent | 23 | PASS |
| **Total Technical Coverage** | **87.5%** | **78-79** | PASS |

### Framework Statistics

| Provider | Total Checks | KSI Categories | Status |
|----------|--------------|----------------|--------|
| **AWS** | 175 | 13 | Validated |
| **Azure** | 186 | 13 | Validated |
| **GCP** | 122 | 13 | Validated |

## Usage Guidance

### Running FedRAMP 20x KSI Low Scans

```bash
# AWS
prowler aws --compliance fedramp_20x_ksi_low_aws

# Azure
prowler azure --compliance fedramp_20x_ksi_low_azure

# GCP
prowler gcp --compliance fedramp_20x_ksi_low_gcp
```

### Interpreting Results

**PASS Results:**
- Technical control is properly implemented
- Configuration meets FedRAMP KSI requirement
- No additional action needed for this specific check

**FAIL Results:**
- Technical control is missing or misconfigured
- Remediation required to meet FedRAMP KSI requirement
- Review check documentation for remediation guidance

**Important:** A complete FedRAMP Low authorization requires **both**:
1. Passing technical checks (automated via Prowler)
2. Organizational controls and documentation (manual validation)

### What Additional Validation Is Required

Organizations must **manually validate** the following to achieve full RFC-0014 Phase Two compliance:

1. **Training & Education (KSI-CED-01, KSI-CED-03):**
   - Security awareness training completion rates
   - Role-specific training curricula
   - Testing and verification records
   - Annual training requirements

2. **CI/CD Pipeline Security (KSI-CMT-03 gap):**
   - Build pipeline security configurations
   - Artifact integrity validation
   - Secret management in build processes
   - Container scanning integration

3. **Policy Documentation:**
   - Incident response procedures
   - Change management processes
   - Access control policies
   - Data classification standards

4. **Third-Party Management (KSI-TPR):**
   - Vendor security assessments
   - Supply chain risk management
   - Contract security requirements
   - Third-party access controls

## Complementary Controls

To achieve full FedRAMP Low authorization, organizations should:

### 1. Document Organizational Policies
- Security awareness training program
- Role-based training requirements
- Change management procedures
- Incident response plans
- Business continuity and disaster recovery

### 2. Implement Manual Reviews
- Quarterly access reviews
- Annual training completion audits
- Third-party security assessments
- Configuration management database validation

### 3. Maintain Evidence
- Training completion records
- Incident response logs and reports
- Change management tickets
- Security assessment reports
- Continuous monitoring reports

### 4. Integrate with Other Tools
- SIEM for security event correlation
- SOAR for incident response automation
- Ticketing systems for change management
- Training management systems for compliance tracking

## Best Practices

1. **Run Scans Regularly:** Schedule automated Prowler scans (daily or weekly) to maintain continuous compliance monitoring
2. **Prioritize Failures:** Focus on FAIL results that impact multiple KSIs first
3. **Document Exceptions:** For technical controls that cannot be implemented, document compensating controls
4. **Combine with Manual Reviews:** Use Prowler results as input to your System Security Plan (SSP)
5. **Track Trends:** Monitor compliance scores over time to demonstrate continuous improvement
6. **Integrate into CI/CD:** Run Prowler checks in deployment pipelines to catch misconfigurations early

## Support & Updates

**Framework Maintenance:**
- This framework is maintained as part of the open-source Prowler project
- Updates will be published to align with FedRAMP RFC changes
- Community contributions are welcome

**Getting Help:**
- GitHub Issues: https://github.com/prowler-cloud/prowler/issues
- Documentation: https://docs.prowler.com/
- FedRAMP Resources: https://www.fedramp.gov/

**Official FedRAMP References:**
- RFC-0014 Phase Two: https://www.fedramp.gov/rfcs/0014/
- KSI Documentation: https://github.com/FedRAMP/docs/blob/main/markdown/FRMR.KSI.key-security-indicators.md

---

## Summary

The FedRAMP 20x KSI Low compliance frameworks provide **strong automated coverage** (87.5%) of RFC-0014 Phase Two technical requirements for Low impact systems. The frameworks excel at validating cloud infrastructure security controls but require complementary organizational policies and manual validation to achieve complete FedRAMP compliance.

**Key Takeaways:**
- Excellent technical control coverage across 483+ checks
- Multi-cloud support (AWS, Azure, GCP)
- Requires manual validation for training/education KSIs
- Requires CI/CD pipeline security validation if using AWS-native tools
- Provides continuous monitoring capability
- Integrates into existing compliance workflows

**Recommended Use:** Deploy these frameworks as part of a comprehensive FedRAMP compliance program that includes both automated technical assessments (Prowler) and manual organizational control validation.

ethanolivertroy commented:
Notes-

# FedRAMP 20x KSI Low Compliance Framework - Limitations & Usage Guide

## Overview

The FedRAMP 20x Key Security Indicators (KSI) Low Impact Level compliance frameworks provide automated security assessment for cloud service providers pursuing FedRAMP Low authorization. These frameworks implement [RFC-0014 Phase Two](https://www.fedramp.gov/rfcs/0014/) requirements across AWS, Azure, and GCP.

## Scope & Purpose

**What This Framework Checks:**
- Technical cloud infrastructure security controls
- Identity and access management configurations
- Encryption implementation (at rest and in transit)
- Network segmentation and traffic controls
- Logging, monitoring, and audit configurations
- Service configuration security settings
- Change management through automated tooling
- High availability and resilience architecture

**What This Framework Does NOT Check:**
- Organizational policies and procedures
- Employee training and awareness programs
- Physical security controls
- Manual documentation reviews
- Contractual agreements with third parties
- Business continuity plan documents
- Incident response runbooks (documents vs. technical controls)

## Known Limitations

### 1. Organizational & Training KSIs Not Covered

The following RFC-0014 Phase Two KSIs are **intentionally excluded** because they require organizational policy validation rather than technical infrastructure checks:

- **KSI-CED-01** (Security & Privacy Awareness Training): Requires evidence of employee training programs, completion rates, and testing results
- **KSI-CED-03** (Role-Specific Security Training): Requires documentation of role-based training curricula and completion tracking

**Rationale:** Prowler focuses on automated technical security assessments of cloud infrastructure. These education requirements must be validated through policy document review, training management systems, and manual audits.

### 2. Technical Implementation Approach

This framework uses an **umbrella category approach** to organize checks:

| Umbrella Category | RFC-0014 Phase Two Requirements Covered |
|-------------------|----------------------------------------|
| **KSI-CMT** (Change Management) | CMT-01 (Service Modification Logging), CMT-04 (Change Management Compliance) |
| **KSI-CNA** (Cloud Native Architecture) | CNA-01 (Network Traffic Limitation) |
| **KSI-IAM** (Identity & Access Management) | IAM-05 (Zero Trust Design) |
| **KSI-SVC** (Service Configuration) | SVC-03 (Encryption at Rest) |
| **KSI-IAM-07** (Account Lifecycle) | Explicitly implemented as dedicated requirement |
| **KSI-MLA-07** (Monitoring Inventory) | Explicitly implemented as dedicated requirement |

**Benefits:** Comprehensive multi-layered coverage, easier maintenance across 483+ checks, and practical CSP validation.

### 3. CI/CD Pipeline Security Gap

**RFC-0014 Requirement:** KSI-CMT-03 (Persistent Automated Testing)

**Current Coverage:** Partial
- Automated compliance monitoring (AWS Config, SSM)
- Automated patch management
- **Gap:** CI/CD pipeline security checks

**Missing Technical Checks:**
- CodeBuild project security configurations
- CodePipeline artifact encryption
- Build-time secret management
- Container image scanning in pipelines

**Why This Gap Exists:**
- Many CSPs use external CI/CD tools (Jenkins, GitLab CI, GitHub Actions) rather than AWS-native services
- The framework focuses on deployed infrastructure rather than build pipelines

**Mitigation:** Organizations using AWS CodeBuild/CodePipeline should manually validate artifact encryption, build logging, secret management, and least privilege IAM roles.

### 4. Subject to RFC-0014 Finalization

**Current Implementation:** Based on RFC-0014 Phase Two as published at https://www.fedramp.gov/rfcs/0014/

**Notes:**
- This framework implements RFC-0014 Phase Two as of November 2025
- Framework will be updated if FedRAMP publishes material changes
- Moderate impact level KSIs (CNA-08, MLA-08, SVC-08, SVC-09, SVC-10) are not included in this Low framework

## Implementation Notes

### What This Framework Implements

**Original KSI Categories (9)**:
- KSI-CED (Cybersecurity Education), KSI-CMT (Change Management), KSI-CNA (Cloud Native Architecture), KSI-IAM (Identity & Access Management), KSI-INR (Incident Response), KSI-MLA (Monitoring, Logging, and Auditing), KSI-PIY (Policy and Inventory), KSI-RPL (Recovery Planning), KSI-SVC (Service Configuration), KSI-TPR (Third-Party Resources)

**RFC-0014 Phase Two Additions for Low (3)**:
- **KSI-CED-03**: Role-Specific Training
- **KSI-IAM-07**: Account Lifecycle Management
- **KSI-MLA-07**: Monitoring and Logging Inventory

**Total**: 12 KSI requirements with 184+ automated security checks

### Evolution from Original KSI to RFC-0014 Phase Two

**Retired KSIs** (consolidated into other requirements):
- KSI-MLA-04 (authenticated vulnerability scanning) - merged into MLA-03
- KSI-MLA-06 (centralized vulnerability tracking) - merged into MLA-03
- KSI-TPR-02 (third-party authorization confirmation) - merged into TPR-01

**Enhanced Requirements**:
- **KSI-INR**: Renamed from "Incident Reporting" to "Incident Response" (reactive to proactive)
- **KSI-CMT-03**: Changed from "pre-deployment" to "persistent" automated testing
- **KSI-SVC-03**: Encryption now required "by default" (expanded scope)
- 19 additional KSIs enhanced with stronger validation requirements

**Philosophical Shift**:
- **Phase One**: Written attestation acceptable for Low impact level
- **Phase Two**: Automated, evidence-based validation with continuous monitoring

### What's Excluded

**Moderate-Only KSIs** (not applicable to Low):
- KSI-CNA-08 (automated security posture assessment), KSI-MLA-08 (least-privilege log access), KSI-SVC-08 (change validation), KSI-SVC-09 (communication integrity), KSI-SVC-10 (data removal)

**Retired KSIs** (superseded by Phase Two consolidation):
- KSI-MLA-04, KSI-MLA-06, KSI-TPR-02

## Coverage Analysis

### RFC-0014 Phase Two Low Impact Level Requirements

| Requirement | Coverage | Check Count | Status |
|-------------|----------|-------------|--------|
| KSI-CMT-01 (Service Modification Logging) | Excellent | 15 | PASS |
| KSI-CMT-03 (Persistent Automated Testing) | Partial | 2-3 | PARTIAL |
| KSI-CMT-04 (Change Management Compliance) | Good | 6 | PASS |
| KSI-CNA-01 (Network Traffic Limitation) | Excellent | 9 | PASS |
| KSI-IAM-05 (Zero Trust Design) | Excellent | 12 | PASS |
| KSI-IAM-07 (Account Lifecycle Management) | Explicit | 6 | PASS |
| KSI-MLA-07 (Monitoring & Logging Inventory) | Explicit | 5 | PASS |
| KSI-SVC-03 (Encryption at Rest) | Excellent | 23 | PASS |
| **Total Technical Coverage** | **87.5%** | **78-79** | PASS |

### Framework Statistics

| Provider | Total Checks | KSI Categories | Status |
|----------|--------------|----------------|--------|
| **AWS** | 175 | 13 | Validated |
| **Azure** | 186 | 13 | Validated |
| **GCP** | 122 | 13 | Validated |

## Usage Guidance

### Running FedRAMP 20x KSI Low Scans

```bash
# AWS
prowler aws --compliance fedramp_20x_ksi_low_aws

# Azure
prowler azure --compliance fedramp_20x_ksi_low_azure

# GCP
prowler gcp --compliance fedramp_20x_ksi_low_gcp
```

### Interpreting Results

**PASS Results:**
- Technical control is properly implemented
- Configuration meets FedRAMP KSI requirement
- No additional action needed for this specific check

**FAIL Results:**
- Technical control is missing or misconfigured
- Remediation required to meet FedRAMP KSI requirement
- Review check documentation for remediation guidance

**Important:** A complete FedRAMP Low authorization requires **both**:
1. Passing technical checks (automated via Prowler)
2. Organizational controls and documentation (manual validation)

### What Additional Validation Is Required

Organizations must manually validate:

1. **Training & Education (KSI-CED-01, KSI-CED-03)** - Security awareness and role-specific training programs
2. **CI/CD Pipeline Security (KSI-CMT-03 gap)** - Build pipeline security if using AWS CodeBuild/CodePipeline
3. **Policy Documentation** - Incident response, change management, access control policies
4. **Third-Party Management (KSI-TPR)** - Vendor assessments and supply chain risk management

## References

- **Original KSI Document**: https://github.com/FedRAMP/docs/blob/main/markdown/FRMR.KSI.key-security-indicators.md (Base requirements)
- **RFC-0014 Phase Two**: https://www.fedramp.gov/rfcs/0014/ (Updates and enhancements)
- **Prowler Documentation**: https://docs.prowler.com/

---

## Summary

The FedRAMP 20x KSI Low frameworks provide **87.5% automated coverage** of RFC-0014 Phase Two technical requirements across 483+ checks for AWS, Azure, and GCP. The frameworks excel at validating cloud infrastructure security but require complementary organizational policies and manual validation for training/education KSIs and CI/CD pipeline security to achieve complete FedRAMP Low compliance.

pedrooot committed:
Merge branch 'master' into feature/fedramp-20x-ksi-compliance

pedrooot committed:
chore: merge master into feature/fedramp-20x-ksi-compliance

pedrooot committed:
chore: update readme

pedrooot committed:
Merge branch 'master' into feature/fedramp-20x-ksi-compliance

ethanolivertroy commented:
Closing this to clean it up after talking with @pedrooot
